diff --git a/examples/src/main/python/ml/pipeline_example.py b/examples/src/main/python/ml/pipeline_example.py
index e1fab7c..e2b6ee8 100644
--- a/examples/src/main/python/ml/pipeline_example.py
+++ b/examples/src/main/python/ml/pipeline_example.py
@@ -62,8 +62,8 @@ if __name__ == "__main__":
     prediction = model.transform(test)
     selected = prediction.select("id", "text", "probability", "prediction")
     for row in selected.collect():
-        rid, text, prob, prediction = row
-        print("(%d, %s) --> prob=%s, prediction=%f" % (rid, text, str(prob), prediction))
+        rid, text, prob, prediction = row  # type: ignore[no-redef]
+        print("(%d, %s) --> prob=%s, prediction=%f" % (rid, text, str(prob), prediction))  # type: ignore
     # $example off$
 
     spark.stop()
