commit 70b0977203c1872e5a648c6fd8452b4ebe6fc50a (HEAD -> SPYT-393_add_spark_patches)
author: alex-shishkin
date: 2022-10-21T12:21:55+03:00

    SPYT-393 added worker log feature

--- taxi/dmp/spark/spark/core/src/main/scala/org/apache/spark/deploy/history/HistoryServer.scala	(72e29fd0562f55e9000fe547087a795647e7ef42)
+++ taxi/dmp/spark/spark/core/src/main/scala/org/apache/spark/deploy/history/HistoryServer.scala	(70b0977203c1872e5a648c6fd8452b4ebe6fc50a)
@@ -177,6 +177,18 @@ class HistoryServer(
         logError("Logs from YT will not be loaded because class in spyt was not found", e)
     }
 
+    try {
+      val logPage = ScalaClassLoader(getClass.getClassLoader)
+        .loadClass("org.apache.spark.deploy.history.YtLogPage")
+        .getConstructor(classOf[SparkConf]).newInstance(conf).asInstanceOf[WorkerLogPage]
+      attachPage(logPage)
+      attachHandler(createServletHandler("/workerLog",
+        (request: HttpServletRequest) => logPage.renderLog(request), conf))
+    } catch {
+      case e: ClassNotFoundException =>
+        logError("Logs from YT will not be loaded because class in spyt was not found", e)
+    }
+
     addStaticHandler(SparkUI.STATIC_RESOURCE_DIR)
 
     val contextHandler = new ServletContextHandler
--- taxi/dmp/spark/spark/core/src/main/scala/org/apache/spark/deploy/history/WorkerLogPage.scala	(72e29fd0562f55e9000fe547087a795647e7ef42)
+++ taxi/dmp/spark/spark/core/src/main/scala/org/apache/spark/deploy/history/WorkerLogPage.scala	(70b0977203c1872e5a648c6fd8452b4ebe6fc50a)
@@ -0,0 +1,33 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.deploy.history
+
+import javax.servlet.http.HttpServletRequest
+
+import scala.xml.Node
+
+import org.apache.spark.SparkConf
+import org.apache.spark.internal.Logging
+import org.apache.spark.ui.WebUIPage
+
+private[history] abstract class WorkerLogPage(conf: SparkConf)
+  extends WebUIPage("workerLogPage") with Logging {
+  def renderLog(request: HttpServletRequest): String
+
+  def render(request: HttpServletRequest): Seq[Node]
+}
